Makefile

```diff
@@ -1,18 +1,56 @@
-all: push
+all: container
 
 # See node-problem-detector.yaml for the version currently running-- bump this ahead before rebuilding!
-TAG = v0.2
+# TAG is the tag of the container image.
+TAG?=v0.2
 
-PROJ = google_containers
+# PROJ is the image project.
+PROJ?=google_containers
 
-node-problem-detector: node_problem_detector.go
-	CGO_ENABLED=0 GOOS=linux godep go build -a -installsuffix cgo -ldflags '-w' -o node-problem-detector
+# ENABLE_JOURNALD enables build journald support or not. Building journald support needs libsystemd-dev
+# or libsystemd-journal-dev.
+# TODO(random-liu): Build NPD inside container.
+ENABLE_JOURNALD?=1
+
+# TODO(random-liu): Support different architectures.
+BASEIMAGE:=alpine:3.1
+
+# Disable cgo by default to make the binary statically linked.
+CGO_ENABLED:=0
+
+# NOTE that enable journald will increase the image size.
+ifeq ($(ENABLE_JOURNALD), 1)
+	# Enable journald bulid tag.
+	BUILD_TAGS:=-tags journald
+	# Use fedora because it has newer systemd version (229) and support +LZ4. +LZ4 is needed
+	# on some os distros such as GCI.
+	BASEIMAGE:=fedora
+	# Enable cgo because sdjournal needs cgo to compile. The binary will be dynamically
+	# linked if CGO_ENABLED is enabled. This is fine because fedora already has necessary
+	# dynamic library. If we need to make this statically linked in the future, add -ldflags
+	# '-extldflags "-static"'
+	CGO_ENABLED:=1
+endif
+
+SOURCES:=$(shell find . -name '*.go')
+
+node-problem-detector: $(SOURCES)
+	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux go build -o node-problem-detector -a -ldflags '-w' $(BUILD_TAGS)
 
 container: node-problem-detector
+	cp Dockerfile.in Dockerfile
+	sed -i 's|BASEIMAGE|$(BASEIMAGE)|g' Dockerfile
+	sed -i 's|SETUP|$(SETUP)|g' Dockerfile
 	docker build -t gcr.io/$(PROJ)/node-problem-detector:$(TAG) .
 
 push: container
 	gcloud docker push gcr.io/$(PROJ)/node-problem-detector:$(TAG)
 
 clean:
+	rm -f Dockerfile
 	rm -f node-problem-detector
+
+test:
+	go test -v -race ./pkg/... $(BUILD_TAGS)
+
+.PHONY: all
```

Comment:
`PHONY` isn't about whether it happens every time, just about whether it treats the target name as a file to check the existence of (e.g. if you do `touch clean; make clean` it won't run the commands because it treats "clean" as a filename and sees that the target already exists).

`node-problem-detector` isn't a PHONY necessarily, but everything else should be.